### Imports

In [1]:
import pandas as pd

|Model|Train Accuracy|Test Accuracy|Correctly Predicted|Incorrectly Preddicted|
|---|---|---|---|---|
|Baseline Spam not Spam model|99.9%|99.7%|1,415|6|
|Baseline Phishing/Malware model|99.4%|74.0%|529|186|
|Model 1 Phishing/Malware model|76.8%|75.4%|539|176|
|Model 2 Phishing/Malware model|82.8%|75.5%|540|175|
|Model 3 Phishing/Malware model|91.0%|73.7%|527|188|
|Model 4 Phishing/Malware model|94.5%|74.97%|536|179|
|Model 5 Phishing/Malware model|70.96%|56.2%|402|313|

To conclude we was able to create our 1st filtering model with 99.7% accuracy on our test data with filtering the Scam data out of the dataset, which ended up being our baseline Logistic Regression model. <br>
<br>
Now for the Phishing and Malware categorizing model as we can see from the data above it is between Model 1 and Model 2. When diving deeper into the split of malware and phising mistakes model 1 had a 138 to 38 split, model 2 had a 127 to 48 split, whereas model 4 has a 94 to 85 split.<br>
<br>
For this situation we are going to focus on balance rather than be biased to 1 particular category of Phishing or Malware, so we will be selecting Model 4 as our final production model in this situation.

### Spam Filtering Model

In [11]:
df = pd.read_csv('../Data/merged_data.csv')
filter_preds = pd.read_csv('../Data/Spam_notSpam_Predictions.csv')

In [12]:
df['Predictions'] = filter_preds['Predictions']
df.head()

created_utc                                              title  \
0  1.697589e+09                  Is my friend about to be scammed?   
1  1.697586e+09  Just Delt with an AirBnB Scam anyone can do an...   
2  1.697586e+09  Spam caller knew my name and 'address' but it ...   
3  1.697587e+09                         I am waiting on a package…   
4  1.697586e+09                  Snapchat Minor Solicitation Scam?   

                                           self_text subreddit  self_text_len  \
0  Friend of mine from back in Uni (before we mov...     Scams            939   
1  Apparently you can instant book seasonal renta...     Scams            481   
2  First off I am new to posting on Reddit so ple...     Scams           1377   
3  So, I got this text message and I’m trying to ...     Scams            192   
4  Added a girl on Snapchat from Omegle. Look at ...     Scams            866   

   self_text_count  punc_count  scam_notscam  Predictions  
0              183          32             1            1  
1               92           4             1            1  
2              254          27             1            1  
3               41           5             1            1  
4              161          24             1            1

In [13]:
conflicts = df.query('scam_notscam != Predictions')
conflicts

created_utc                                              title  \
2797  1.683472e+09  Announcement: Changes, Looking for Mods, Join ...   
2823  1.696004e+09                 So did ANYBODY like Strange world?   
3093  1.669661e+09           What were your thoughts on Bullet Train?   
3475  1.660492e+09                                   Top 20 A24 Films   
3649  1.620843e+09  [Podcast] Not Your Father's Movies | MOTHER’S ...   
3681  1.604949e+09  Cinematic Excess in Sofia Coppola’s Marie Anti...   

                                              self_text     subreddit  \
2797  # This Reddit group is going to have a revamp ...  moviereviews   
2823  Just curious. Pour your heart out. I would giv...  moviereviews   
3093                [Bullet Train: Stupid, Goofy, Fun](  moviereviews   
3475  Ah! Here we go... This list has been a long ti...  moviereviews   
3649  [Website]( //[ Spotify]( //[ Apple Podcasts]( ...  moviereviews   
3681  I recently published an article reviewing Copp...  moviereviews   

      self_text_len  self_text_count  punc_count  scam_notscam  Predictions  
2797            289               53           6             0            1  
2823            122               25           4             0            1  
3093             35                5           3             0            1  
3475            155               31          10             0            1  
3649            464               70           5             0            1  
3681            159               24           2             0            1

Our 1st filtering model had a 99.7% accuracy on our test data with filtering the Scam data out of the dataset, which ended up being our baseline Logistic Regression model. We can see from the table above that the model predicted that 6 of our movie reviews as scams. Now when looking at them you can understand why some of these may have been classified as such with the weird punctuation use and format of the text and then the first one starting of with Announcement, which could have potentially been used on multiple scam messages. 

### Malware or Phishing Filtering

In [14]:
df1 = pd.read_csv('../Data/Phis_orMal_predictions.csv')
df1

created_utc                                              title  \
0     1.697552e+09  Malware on New York Times top sponsored link u...   
1     1.697553e+09         Why would a Url redirect to multiple sites   
2     1.697547e+09                   Accessibility malware on android   
3     1.697298e+09  Benefits of Dynamic analysis vs Any.run(intera...   
4     1.697571e+09                         Opened a spam pdf in hurry   
...            ...                                                ...   
2855  1.682629e+09                                    Kidnapping scam   
2856  1.697448e+09  Just received an SMS verification code from Si...   
2857  1.697560e+09  fake vehicle dealer: aplusautomotive.com "A Pl...   
2858  1.697123e+09  a recruiting company is reaching out to me ask...   
2859  1.697548e+09                                Very elaborate scam   

                                              self_text subreddit  \
0     Today I Googled "NYT" and clicked the top spon...   Malware   
1     I saw gmai.com on my security audit this morni...   Malware   
2     If a user downloads a malicious app from outsi...   Malware   
3     Hello,\n\nI was wondering since [any.run](http...   Malware   
4     Recieved an email on my phone saying norton su...  phishing   
...                                                 ...       ...   
2855  So recently a friend showed up at my house cry...  Phishing   
2856  So I am a little concerned here for a couple o...  Phishing   
2857  Previously "Pro Era Group"\n\n[https://www.red...  Phishing   
2858  I am actually looking for an opportunity to gr...  Phishing   
2859  Anyone ever see something like this.\n\n&#x200...  Phishing   

      self_text_len  self_text_count  punc_count  phis_orMal  \
0               657              111          19           0   
1               325               61           5           0   
2               470               73           5           0   
3               356               50          10           0   
4               380               69           8           1   
...             ...              ...         ...         ...   
2855            522              100           8           1   
2856            708              134          11           1   
2857            313               16          11           1   
2858            365               47          10           1   
2859            163                8           7           1   

                                                 merged  y_pred  
0     Malware on New York Times top sponsored link u...       0  
1     Why would a Url redirect to multiple sites I s...       0  
2     Accessibility malware on android If a user dow...       0  
3     Benefits of Dynamic analysis vs Any.run(intera...       0  
4     Opened a spam pdf in hurry Recieved an email o...       1  
...                                                 ...     ...  
2855  Kidnapping scam So recently a friend showed up...       1  
2856  Just received an SMS verification code from Si...       0  
2857  fake vehicle dealer: aplusautomotive.com "A Pl...       0  
2858  a recruiting company is reaching out to me ask...       1  
2859  Very elaborate scam Anyone ever see something ...       1  

[2860 rows x 10 columns]

In [17]:
conflicts2 = df1.query('phis_orMal != y_pred')
conflicts2

created_utc                                              title  \
18    1.696940e+09                     Duplicated domain. What to do?   
20    1.696799e+09                                    Shortcode 54030   
67    1.479308e+09  Malware spreading through Android devices, via...   
70    1.599608e+09           Countries Seeing Spike in Emotet Attacks   
86    1.553694e+09  The Definitive Guide to Implementing DMARC to ...   
...            ...                                                ...   
2826  1.680481e+09             Art Scam Alert - Quarters on Red Bluff   
2828  1.697012e+09                        Is Good Gamer Group a scam?   
2834  1.697113e+09                     is helloice clothing any good?   
2856  1.697448e+09  Just received an SMS verification code from Si...   
2857  1.697560e+09  fake vehicle dealer: aplusautomotive.com "A Pl...   

                                              self_text subreddit  \
18    Hi. Our website [uvodo.com](https://uvodo.com)...  phishing   
20           Does anyone know who's shortcode 54030 is?  phishing   
67    A friend of mine, via her Android 6.0 device, ...   Malware   
70    [https://www.zdnet.com/article/france-japan-ne...   Malware   
86    As more and more organizations realize that SP...   Malware   
...                                                 ...       ...   
2826  Hi everyone! I had recently seen a posting her...  Phishing   
2828  I signed up to a session of the Good Gamer Gro...  Phishing   
2834  saw an ad and liked some of its clothing, so b...  Phishing   
2856  So I am a little concerned here for a couple o...  Phishing   
2857  Previously "Pro Era Group"\n\n[https://www.red...  Phishing   

      self_text_len  self_text_count  punc_count  phis_orMal  \
18              229               24          15           1   
20               42                7           2           1   
67             4643              526         156           0   
70              840              101          19           0   
86              937              119          23           0   
...             ...              ...         ...         ...   
2826           1963              369          37           1   
2828            127               28           2           1   
2834             93               19           1           1   
2856            708              134          11           1   
2857            313               16          11           1   

                                                 merged  y_pred  
18    Duplicated domain. What to do? Hi. Our website...       0  
20    Shortcode 54030 Does anyone know who's shortco...       0  
67    Malware spreading through Android devices, via...       1  
70    Countries Seeing Spike in Emotet Attacks [http...       1  
86    The Definitive Guide to Implementing DMARC to ...       1  
...                                                 ...     ...  
2826  Art Scam Alert - Quarters on Red Bluff Hi ever...       0  
2828  Is Good Gamer Group a scam? I signed up to a s...       0  
2834  is helloice clothing any good? saw an ad and l...       0  
2856  Just received an SMS verification code from Si...       0  
2857  fake vehicle dealer: aplusautomotive.com "A Pl...       0  

[348 rows x 10 columns]

With Malware and Phishing data being very close together, in regards to the terminology used and that this data is postings of peoples experience rather than the raw data directly from the scammer, it is easy to understand why the model was strugglying with improving it's learning process of the data and making the false predictions.<br>
<br>
In the post that perticularly stands out in the conflicts we can see above is row 3(67) where we can see Malware is specifically mentioned in the title, however it has still been predicted as Phishing. Now this could be due to a couple of reason, first being that if the word has only been mentioned once, the model may have over look the significance of the word. Secondly as the post has been made about someone experience, it could have been miss classifed by the poster and the model was able to determine from the other text that it matches the pattern of a phishing situation.<br>
<br>
Looking at the over performance of we can see that out of 2860 posts 348 posts were falsely predicted giving us an overall accuracy of around 87.83%